spacy/pipeline/entity_linker.py

```diff
@@ -440,96 +458,104 @@ def predict(self, docs: Iterable[Doc]) -> List[str]:
+        valid_ent_idx_per_doc = (
+            [
+                idx
+                for idx in range(len(doc.ents))
+                if doc.ents[idx].label_ not in self.labels_discard
+            ]
+            for doc in docs
+            if len(doc.ents)
+        )
+        # Call candidate generator.
+        if self.candidates_doc_mode:
+            all_ent_cands = self.get_candidates_all(
+                self.kb,
+                (
+                    [doc.ents[idx] for idx in next(valid_ent_idx_per_doc)]
+                    for doc in docs
+                    if len(doc.ents)
+                ),
+            )
+        else:
+            # Alternative: collect entities the old-fashioned way - by retrieving entities individually.
+            all_ent_cands = (
+                [
+                    self.get_candidates(self.kb, doc.ents[idx])
+                    for idx in next(valid_ent_idx_per_doc)
                 ]
+                for doc in docs
+                if len(doc.ents)
+            )
 
-                batch_candidates = list(
-                    self.get_candidates_batch(
-                        self.kb, [ent_batch[idx] for idx in valid_ent_idx]
-                    )
-                    if self.candidates_batch_size > 1
-                    else [
-                        self.get_candidates(self.kb, ent_batch[idx])
-                        for idx in valid_ent_idx
-                    ]
-                )
-
-                # Looping through each entity in batch (TODO: rewrite)
-                for j, ent in enumerate(ent_batch):
-                    sent_index = sentences.index(ent.sent)
-                    assert sent_index >= 0
-
-                    if self.incl_context:
-                        # get n_neighbour sentences, clipped to the length of the document
-                        start_sentence = max(0, sent_index - self.n_sents)
-                        end_sentence = min(
-                            len(sentences) - 1, sent_index + self.n_sents
-                        )
-                        start_token = sentences[start_sentence].start
-                        end_token = sentences[end_sentence].end
-                        sent_doc = doc[start_token:end_token].as_doc()
-                        # currently, the context is the same for each entity in a sentence (should be refined)
-                        sentence_encoding = self.model.predict([sent_doc])[0]
-                        sentence_encoding_t = sentence_encoding.T
-                        sentence_norm = xp.linalg.norm(sentence_encoding_t)
-                    entity_count += 1
-                    if ent.label_ in self.labels_discard:
-                        # ignoring this entity - setting to NIL
+        for doc_idx, doc in enumerate(docs):
+            if len(doc) == 0:
+                continue
+            sentences = [s for s in doc.sents]
+            doc_ent_cands = list(next(all_ent_cands)) if len(doc.ents) else []
```

Comment:
On the topic of empty Docs, is the iteration here OK?

If I understand correctly, `all_ent_cands` has one element per Doc. Those elements (lists) can be empty if a Doc has no entities, but they are always present.

Here the input Docs and `all_ent_cands` are iterated separately. Because of the early `continue` for Docs with no tokens, wouldn't that result in a misalignment, because the `all_ent_cands` generator wouldn't be advanced?

Reply:
So I'll added a [test to check for this](https://github.com/explosion/spaCy/pull/11669/files#diff-60c79e322819366d1dcdbec89602b0c008dc4b8f15832da6ca6ffe8ebf666fb6R1214), but I'm running into the problem that motivated https://github.com/explosion/spaCy/pull/11640 - perhaps we can merge that that?

Reply:
#11640 has been merged, so I guess this should be OK now once that's incorporated?

Reply:
Just merged - tests succeed locally.